### PR TITLE
chore: port over PyTorch example to use Trainer API [MLG-1181]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

since the introduction of Trainer API in 0.21.0, harness codepaths are now considered “legacy” for PyTorchTrial. this ports over existing PyTorch examples (only one now after examples pruning) in the examples repository to showcase the new Trainer API codepath.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: examples/tutorials/mnist_pytorch/data.py [hunk 1]
```diff
@@ -1,46 +1,27 @@
...
-    shutil.unpack_archive(filepath, download_directory)
-
-    return os.path.dirname(download_directory)
+    data_path = pathlib.Path(data_dir)
+    data_path.mkdir(parents=True, exist_ok=True)
+
+    # Use a file lock so that only one worker on each node downloads.
+    with filelock.FileLock(str(data_path / "lock")):
+        return datasets.MNIST(
+            root=str(data_dir),
```
Comment (rb-determined-ai) [inline examples/tutorials/mnist_pytorch/data.py:15]: already a str

The number of str -> pathlib -> str conversions here could use some reconsideration

File: examples/tutorials/mnist_pytorch/data.py
Comment (rb-determined-ai) [inline examples/tutorials/mnist_pytorch/data.py:1]: my take is that this doesn't prolly need its own file anymore, but up to you

File: examples/tutorials/mnist_pytorch/adaptive.yaml [hunk 2]
```diff
@@ -29,5 +26,5 @@ searcher:
   smaller_is_better: true
   max_trials: 16
   max_length:
-    batches: 937 #60,000 training images with batch size 64
```
Comment (rb-determined-ai) [inline examples/tutorials/mnist_pytorch/adaptive.yaml:32]: pretty sure that asha + max length of 1 is going to behave poorly.  The master doesn't do fractions, so asha will just emit everything with epoch length of 1.

Use batches here instead, I think.

File: examples/tutorials/mnist_pytorch/README.md [hunk 3]
```diff
@@ -1,27 +1,46 @@
...
+then run `python3 train.py` from your local environment.
+
+### On-cluster
+To run training on-cluster, configure the desired training arguments (checkpoint or validation periods, 
+checkpoint to start from, etc.) in the `Trainer.fit()` call. An experiment configuration file is also required 
+for on-cluster experiments (several examples are included in the directory).
+
+Then the code can be submitted to Determined for on-cluster training by running this command from the current directory:
+`det experiment create const.yaml .`. The other configurations can be run by specifying the desired 
 configuration file in place of `const.yaml`.
```
Comment (rb-determined-ai) [inline examples/tutorials/mnist_pytorch/README.md:45]: Since you are steering people towards torch.distributed (a decision I agree with), you might want to document what entrypoint changes are needed there.

Also, I don't actually know: does det.launch.torch_distributed work with slots_per_trial=1?

File: examples/tutorials/mnist_pytorch/data.py (hunk 1 above)
Comment (azhou-determined) [reply examples/tutorials/mnist_pytorch/data.py:14]: passing in a pathlib.Path now... not much of an improvement, but looks slightly better.

File: examples/tutorials/mnist_pytorch/data.py
Comment (azhou-determined) [reply examples/tutorials/mnist_pytorch/data.py:1]: i kept it separate to demonstrate separating out the different parts of your workflow. but yeah in this example it's rather useless.

File: examples/tutorials/mnist_pytorch/README.md (hunk 3 above)
Comment (azhou-determined) [reply examples/tutorials/mnist_pytorch/README.md:45]: added a section for experiment config changes. 

> det.launch.torch_distributed work with slots_per_trial=1?
yeah, it does. don't really want to put it in the `const.yaml` example though cause it seems like a waste to start torch distributed if it's not necessary.

Comment (rb-determined-ai) [reply examples/tutorials/mnist_pytorch/README.md:45]: At least document that ddp works with a single slot too. "can I have one entrypoint to rule them all" is like an immediate question for the end user, I think.

**Review by rb-determined-ai — APPROVED** [review]
